# HW8: Vector, matrix, covariance, and correlation coefficient

Total : 50 pts This is an individual assignment.

The first **three** problems are analytical and must be solved by hand. 

The last problem is python programming. **Submit your solution in both handwritten part and jupyter notebook and pdf print of the jupyter notebook.**

# Problem 1 - Vector Projections
(15 pt)
Answer the following problems: 

1. (5 pt) Compute the cosine similarity between these two vectors $\mathbf{b}=\left[\begin{matrix}1\\1\\1\end{matrix}\right]$ and $\mathbf{a}=\left[\begin{matrix}1\\2\\2\end{matrix}\right]$. 

2. (5 pt)Project the vector $\mathbf{b}$ onto $\mathbf{a}$. Check that $\mathbf{e}$ is perpendicular to $\mathbf{a}$, that is the angle between $\mathbf{e}$ and $\mathbf{a}$ is $\pm \pi/2$.

    1. \begin{align*}\mathbf{b}=\left[\begin{matrix}1\\2\\2\end{matrix}\right]~~~~~~\text{and }~~~~~~~\mathbf{a}=\left[\begin{matrix}1\\1\\1\end{matrix}\right]\end{align*}
    2. \begin{align*}\mathbf{b}=\left[\begin{matrix}1\\3\\1\end{matrix}\right]~~~~~~\text{and }~~~~~~~\mathbf{a}=\left[\begin{matrix}-1\\-3\\-1\end{matrix}\right]\end{align*}
    

3. (5 pt) Draw the projection of $\mathbf{b}$ onto $\mathbf{a}$ and also compute the projection vector:
    1. \begin{align*}\mathbf{b}=\left[\begin{matrix}\cos\theta\\\sin\theta\end{matrix}\right]~~~~~~\text{and }~~~~~~~\mathbf{a}=\left[\begin{matrix}1\\0\end{matrix}\right]\end{align*}
    
    hint: $\theta$ is a variable. your answer will be a function of $\theta$
    
    2. \begin{align*}
\mathbf{b}=\left[\begin{matrix}1\\1\end{matrix}\right] ~~~~~~\text{and }~~~~~~~
\mathbf{a}=\left[\begin{matrix}1\\-1\end{matrix}\right]
\end{align*}

# Solution:

1. $\cos\theta = \frac{\mathbf{b} \cdot \mathbf{a}}{\|\mathbf{b}\| \|\mathbf{a}\|} = \frac{1\times1+1\times2+1\times2}{\sqrt{1^2+1^2+1^2}\times\sqrt{1^2+2^2+2^2}} = \frac{5}{\sqrt{3}\times\sqrt{9}} = \frac{5}{3\sqrt{3}}$

2. $proj_\mathbf{a} \mathbf{b} = \mathbf{p} = \left( \frac{\mathbf{b} \cdot \mathbf{a}}{\|\mathbf{a}\|^2} \right) \mathbf{a}, \mathbf{e} = \mathbf{b} - \mathbf{p}$

  (1) $\mathbf{p} = \frac{1\times1+2\times1+2\times1}{1^2+1^2+1^2} \left[\begin{matrix}1\\1\\1\end{matrix}\right] = \left[\begin{matrix}\frac{5}{3}\\\frac{5}{3}\\\frac{5}{3}\end{matrix}\right], \\ \mathbf{e}=\left[\begin{matrix}1\\2\\2\end{matrix}\right] - \left[\begin{matrix}\frac{5}{3}\\\frac{5}{3}\\\frac{5}{3}\end{matrix}\right] = \left[\begin{matrix}-\frac{2}{3}\\\frac{1}{3}\\\frac{1}{3}\end{matrix}\right],\\ \mathbf{e} \cdot \mathbf{a} = -\frac{2}{3} \times 1 + \frac{1}{3} \times 1 + \frac{1}{3} \times 1 = 0$

  (2) $\mathbf{p} = \frac{1\times(-1)+3\times(-3)+1\times(-1)}{(-1)^2+(-3)^2+(-1)^2} \left[\begin{matrix}-1\\-3\\-1\end{matrix}\right] = \left[\begin{matrix}1\\3\\1\end{matrix}\right], \\ \mathbf{e}=\left[\begin{matrix}1\\3\\1\end{matrix}\right] - \left[\begin{matrix}1\\3\\1\end{matrix}\right] = \left[\begin{matrix}0\\0\\0\end{matrix}\right],\\ \mathbf{e} \cdot \mathbf{a} = 0$

3. 

  (1) $\mathbf{p} = \frac{\cos\theta\times1+\sin\theta\times0}{1^2+0^2} \left[\begin{matrix}1\\0\end{matrix}\right] = \left[\begin{matrix}\cos\theta\\0\end{matrix}\right]$

  (2) $\mathbf{p} = \frac{1\times1+1\times(-1)}{1^2+(-1)^2} \left[\begin{matrix}1\\-1\end{matrix}\right] = \left[\begin{matrix}0\\0\end{matrix}\right]$

# Problem 2

(each 2pt, total 10pt)
Let $\mathbf{A}$ be an $n\times m$ matrix, $\mathbf{x}$ be an $m\times 1$ vector, and $\mathbf{B}$ be an $n \times n$ matrix.

For the following operations, state where the operation is valid and, if it is, report the size of the output:

a) $\mathbf{A}^T\mathbf{x}$

b) $\mathbf{A}\mathbf{x}$

c) $\mathbf{B}^T\mathbf{A}\mathbf{x}$

d) $\mathbf{A}\mathbf{B}^T\mathbf{x}$

e) $\mathbf{A}\mathbf{x} + \mathbf{A}\mathbf{B}\mathbf{x}$

# Solution:

a) invalid, $(m \times n) \times (m \times 1)$

b) valid, $(n \times m) \times (m \times 1) = (n \times 1)$

c) valid, $(n \times n) \times (n \times m) \times (m \times 1) = (n \times 1)$

d) invalid, $(n \times m) \times (n \times n) \times (m \times 1)$

e) invalid, $((n \times m) \times (m \times 1)) + ((n \times m) \times (n \times n) \times (m \times 1))$


# Problem 3
(10pt)

Multiply these matrices:

1. \begin{align*}
\left[\begin{matrix}0&0&1\\0&1&0\\1&0&0\end{matrix}\right]\left[\begin{matrix}1&2&3\\4&5&6\\7&8&9\end{matrix}\right]\left[\begin{matrix}0&0&1\\0&1&0\\1&0&0\end{matrix}\right]
\end{align*}
multiply the first two and then use the answer to multiple with the last matrix.

2. \begin{align*}
\left[\begin{matrix}1&0&0\\-1&1&0\\-1&0&1\end{matrix}\right]\left[\begin{matrix}1&2&3\\1&3&1\\1&4&0\end{matrix}\right]
\end{align*}

# Solution

1. \begin{align*}
\left[\begin{matrix}0&0&1\\0&1&0\\1&0&0\end{matrix}\right]\left[\begin{matrix}1&2&3\\4&5&6\\7&8&9\end{matrix}\right]\left[\begin{matrix}0&0&1\\0&1&0\\1&0&0\end{matrix}\right] =\\
\left[\begin{matrix}7&8&9\\4&5&6\\1&2&3\end{matrix}\right]\left[\begin{matrix}0&0&1\\0&1&0\\1&0&0\end{matrix}\right] = \left[\begin{matrix}9&8&7\\6&5&4\\3&2&1\end{matrix}\right]
\end{align*}

2. \begin{align*}
\left[\begin{matrix}1&0&0\\-1&1&0\\-1&0&1\end{matrix}\right]\left[\begin{matrix}1&2&3\\1&3&1\\1&4&0\end{matrix}\right] = \left[\begin{matrix}1&2&3\\0&1&-2\\0&2&-3\end{matrix}\right]
\end{align*}

# Problem 4

(15pt)

This following dataset contains information about individuals’ medical insurance charges along with attributes such as age, sex, body mass index (BMI), number of children, smoking status, and region.

https://www.kaggle.com/datasets/mirichoi0218/insurance


we are interested in exploring how these different attributes correlate with medical insurance charges of the individual. 

1. (5pt) consider the individual living in northwest region who does not have kids, which features among bmi, sex, smoker, regions are highly correlated with the insurance cost? 

(note the smoker and sex are not real valued. You can covert them to real valued entries using Boolean, for example, True for male and False for female, True for smoker and False for nonsmoker)
Without the conversion, these data entry will be ignored in dataframe cov and corr functions.


In [1]:
import pandas as pd
import numpy as np

df= pd.read_csv( 'insurance.csv' )

df['sex_num'] = df['sex'].map({'male': 1, 'female': 0})
df['smoker_num'] = df['smoker'].map({'yes': 1, 'no': 0})

northwest_no_kids = df[(df['region'] == 'northwest') & (df['children'] == 0)]
northwest_no_kids

age     sex     bmi  children smoker     region      charges  sex_num  \
3      33    male  22.705         0     no  northwest  21984.47061        1   
4      32    male  28.880         0     no  northwest   3866.85520        1   
9      60  female  25.840         0     no  northwest  28923.13692        0   
33     63    male  28.310         0     no  northwest  13770.09790        1   
35     19    male  20.425         0     no  northwest   1625.43375        1   
...   ...     ...     ...       ...    ...        ...          ...      ...   
1237   58  female  28.215         0     no  northwest  12224.35085        0   
1271   25  female  34.485         0     no  northwest   3021.80915        0   
1277   32  female  29.735         0     no  northwest   4357.04365        0   
1311   33  female  26.695         0     no  northwest   4571.41305        0   
1337   61  female  29.070         0    yes  northwest  29141.36030        0   

      smoker_num  
3              0  
4              0  
9              0  
33             0  
35             0  
...          ...  
1237           0  
1271           0  
1277           0  
1311           0  
1337           1  

[132 rows x 9 columns]

In [2]:
correlations_q1 = {}

# BMI correlation with charges
bmi_corr = np.corrcoef(northwest_no_kids['bmi'], northwest_no_kids['charges'])[0, 1]
correlations_q1['bmi'] = bmi_corr

# Sex correlation with charges
sex_corr = np.corrcoef(northwest_no_kids['sex_num'], northwest_no_kids['charges'])[0, 1]
correlations_q1['sex_num'] = sex_corr

# Smoker correlation with charges
smoker_corr = np.corrcoef(northwest_no_kids['smoker_num'], northwest_no_kids['charges'])[0, 1]
correlations_q1['smoker_num'] = smoker_corr

# Age correlation with charges
age_corr = np.corrcoef(northwest_no_kids['age'], northwest_no_kids['charges'])[0, 1]
correlations_q1['age'] = age_corr

print("Question 1 - Correlations with charges for northwest region with no children:")
for feature, corr in correlations_q1.items():
    print(f"{feature}: {corr:.4f}")

Question 1 - Correlations with charges for northwest region with no children:
bmi: 0.1336
sex_num: -0.1313
smoker_num: 0.7628
age: 0.3218


2. (5pt) From your analysis, which variables have negative correlation to the charges in the northwest region?     Explain what  these negative correlations imply with your own understanding.


In [3]:
print("\nQuestion 2 - Variables with negative correlation:")
for feature, corr in correlations_q1.items():
    if corr < 0:
        print(f"{feature}: {corr:.4f}")


Question 2 - Variables with negative correlation:
sex_num: -0.1313


Since sex_num was mapped as 1 for males and 0 for females, this negative correlation implies that as the value of sex_num decreases (moving from male to female), the insurance charges tend to increase. In other words, females in the northwest region with no children tend to have higher insurance charges than males in the same demographic group.

3. (5 pt) Now let's convert the regions to real numbers: 0: northwest, 1: northeast, 2: southwest, 3: southeast. Perform the correlation analysis and explain if the charges are correlated with the region?

In [4]:
df= pd.read_csv( 'insurance.csv' )
region_mapping = {
    'northwest': 0,
    'northeast': 1,
    'southwest': 2,
    'southeast': 3
}

df['region_num'] = df['region'].map(region_mapping)

In [5]:
region_corr = np.corrcoef(df['region_num'], df['charges'])[0, 1]

print("\nQuestion 3 - Correlation between region and charges:")
print(f"Region correlation: {region_corr:.4f}")

region_avg = df.groupby('region')['charges'].mean().reset_index()

print("\nAverage charges by region:")
for index, row in region_avg.iterrows():
    print(f"{row['region']}: ${row['charges']:.2f}")


Question 3 - Correlation between region and charges:
Region correlation: 0.0570

Average charges by region:
northeast: $13406.38
northwest: $12417.58
southeast: $14735.41
southwest: $12346.94


The correlation between region (coded as 0: northwest, 1: northeast, 2: southwest, 3: southeast) and charges is 0.0570, which indicates a very weak positive correlation.
This weak positive correlation suggests that as the region code increases (moving from northwest to southeast), there's a slight tendency for insurance charges to increase as well, but this relationship is minimal. A correlation coefficient of 0.0570 is so close to zero that the relationship is nearly negligible.